src/js/integrations/devicecontext.ts

```diff
@@ -19,26 +21,79 @@ export class DeviceContext implements Integration {
   /**
    * @inheritDoc
    */
-  public setupOnce(): void {
+  public setupOnce(
+    addGlobalEventProcessor: (callback: EventProcessor) => void,
+    getCurrentHub: () => Hub,
+  ): void {
     addGlobalEventProcessor(async (event: Event) => {
       const self = getCurrentHub().getIntegration(DeviceContext);
       if (!self) {
         return event;
       }
 
+      let contexts: NativeDeviceContextsResponse | null = null;
       try {
-        const contexts = await NATIVE.fetchNativeDeviceContexts();
+        contexts = await NATIVE.fetchNativeDeviceContexts();
+      } catch (e) {
+        logger.log(`Failed to get device context from native: ${e}`);
+      }
+
+      if (!contexts) {
+        return event;
+      }
+
+      const nativeUser = contexts['user'] as Event['user'];
+      if (!event.user && nativeUser) {
+        event.user = nativeUser;
+      }
+
+      const nativeContext = contexts['context'] as Event['contexts'];
+      if (nativeContext) {
+        event.contexts = { ...nativeContext, ...event.contexts };
+      }
 
-        const context = contexts['context'] as Contexts ?? {};
-        const user = contexts['user'] ?? {};
+      const nativeTags = contexts['tags'] as Event['tags'];
+      if (nativeTags) {
+        event.tags = { ...nativeTags, ...event.tags };
+      }
+
+      const nativeExtra = contexts['extra'] as Event['extra'];
+      if (nativeExtra) {
+        event.extra = { ...nativeExtra, ...event.extra };
+      }
+
+      const nativeFingerprint = contexts['fingerprint'] as Event['fingerprint'];
+      if (nativeFingerprint) {
+        event.fingerprint = (event.fingerprint ?? []).concat(
+          nativeFingerprint.filter((item) => (event.fingerprint ?? []).indexOf(item) < 0),
+        )
+      }
 
-        event.contexts = { ...context, ...event.contexts };
+      const nativeLevel = typeof contexts['level'] === 'string'
+        ? severityLevelFromString(contexts['level'])
+        : undefined;
+      if (!event.level && nativeLevel) {
+        event.level = nativeLevel;
+      }
+
+      const nativeEnvironment = contexts['environment'] as Event['environment'];
+      if (!event.environment && nativeEnvironment) {
+        event.environment = nativeEnvironment;
+      }
 
-        if (!event.user) {
-          event.user = { ...user };
+      const nativeBreadcrumbs = Array.isArray(contexts['breadcrumbs'])
+        ? contexts['breadcrumbs'].map(breadcrumbFromObject)
+        : undefined;
+      if (nativeBreadcrumbs) {
+        event.breadcrumbs = event.breadcrumbs ?? []
+        for (const breadcrumb of nativeBreadcrumbs ?? []) {
+          const equals = (b: Breadcrumb): boolean => breadcrumbEquals(b, breadcrumb);
```

Comment:
Yes, it's duplication.

Reply:
How do we avoid the dupe on Android? this has a runtime cost, wondering if we can make it better/align the solution across Android/iOS.

Reply:
Oh, I did not think about that.
Maybe we can just remove the JS breadcrumbs and use the iOS ones since all the JS breadcrumbs are synced down.
On android we do the same, we remove the JS breadcrumbs and use Android only.